[CORE-13776] Handle `rpk.describe_group` racing with `__consumer_offsets` creation

dotnwat opened the pull request:
```
Module:      rptest.tests.log_segment_ms_test
Class:       SegmentMsTest
Method:      test_segment_rolling_with_retention_consumer
Nodes (used/allocated): 4/4

  File "/root/tests/rptest/clients/rpk.py", line 1047, in try_describe_group
    coordinator_partition = parse_field("COORDINATOR-PARTITION", lines[2])
  File "/root/tests/rptest/clients/rpk.py", line 1015, in parse_field
    assert m is not None, f"Field string '{string}' does not match the pattern"
AssertionError: Field string 'STATE             Dead' does not match the pattern
```

Here is normal output from rpk when parsing

```
[INFO  - 2025-11-19 02:33:29,103 - rpk - try_describe_group - lineno:1056]: 
GROUP                  test-group
COORDINATOR-NODE       1
COORDINATOR-PARTITION  __consumer_offsets/8
STATE                  Empty
BALANCER               
MEMBERS                0
TOTAL-LAG              0
```

Here it is when we see this exception

```
[DEBUG - 2025-11-17 06:56:35,198 - rpk - _execute - lineno:1482]: 
GROUP             test-group
COORDINATOR-NODE  1
STATE             Dead
BALANCER          
MEMBERS           0
TOTAL-LAG         0
```

Indeed, that is the case in rpk

```
	if partitionCount > 0 {
		// We do calculate the partition ID for the group with the following
		// hash-based algorithm, which is Redpanda specific at least at 25.1.x, https://github.com/redpanda-data/redpanda/blob/v25.1.1/src/v/kafka/server/coordinator_ntp_mapper.h#L50-L60
		// Apache Kafka does use a different algorithm, https://github.com/apache/kafka/blob/4.0.0/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L192
		hash := xxhash.Sum64String(group.Group)
		partitionID := jump.Hash(hash, int32(partitionCount))
		fmt.Fprintf(tw, "COORDINATOR-PARTITION\t__consumer_offsets/%d\n", partitionID)
	}
```

So presumably rpk is conditionally logging the coordinator partition field.

We can see this in the logs

```
A) here is where rpk describe is begin called by ducktape. one of the very first things that it does is checks how many partitions the consumer offsets topic has.

[DEBUG - 2025-11-17 06:56:34,631 - rpk - _execute - lineno:1455]: Executing command: ['/var/lib/buildkite-agent/builds/redpanda-bk-agent-v6-core-m5d12xlarge-xfs-partitions-i-0e274305d11b8b15c-1/redpanda/redpanda/vbuild/redpanda_installs/ci/bin/rpk', 'group', '-X', 'brokers=docker-rp-16:9092', 'describe', 'test-group', '-X', 'globals.request_timeout_overhead=30s', '-v']

B) here is that metadata call about 50ms later

TRACE 2025-11-17 06:56:34,684 [shard 0:kafk] kafka - handler.h:81 - [client_id: {rpk}] handling metadata v12 request {topics={[{topic_id=AAAAAAAAAAAAAAAAAAAAAA== name={__consumer_offsets}}]} allow_auto_topic_creation=false include_cluster_authorized_operations=false include_topic_authorized_operations=false}
DEBUG 2025-11-17 06:56:34,684 [shard 0:kafk] kafka - request_context.h:238 -
[192.168.215.28:51426] sending 3:metadata for {rpk}, response
{throttle_time_ms=0ms brokers={node_id=1 host=docker-rp-16 port=9092
rack={nullopt}} cluster_id={redpanda.21650b5b-9fbf-43ac-88ae-05ee7363042a}
controller_id=1 topics={error_code={ error_code: unknown_topic_or_partition [3]
} name={__consumer_offsets} topic_id=AAAAAAAAAAAAAAAAAAAAAA== is_internal=false
partitions= topic_authorized_operations=-2147483648}
cluster_authorized_operations=-2147483648}

C) above we see that the topic doesn't exist. rpk will handle this later by using partition count = 0 which alters the output causing the parsing issue.

D) then we see that the consumer offsets topic is created a second or so later.

TRACE 2025-11-17 06:56:34,685 [shard 1:kafk] kafka - requests.cc:96 - [192.168.215.28:51432] processing name:find_coordinator, key:10, version:4 for rpk, mem_units: 8056, ctx_size: 14
TRACE 2025-11-17 06:56:34,685 [shard 1:kafk] kafka - handler.h:81 - [client_id: {rpk}] handling find_coordinator v4 request {key= key_type={group} coordinator_keys=[test-group]}


TRACE 2025-11-17 06:56:34,686 [shard 1:kafk] kafka - group_initializer.cc:37 - can't find consumer group topic, creating
TRACE 2025-11-17 06:56:34,686 [shard 1:kafk] cluster - topics_frontend.cc:1208 - Auto create topics [{ topic: {kafka/__consumer_offsets}, partition_count: 16, replication_factor: 1, is_migrated: true, topic_id: {nullopt}, properties: { compression: {nullopt}, 
```

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none


[CORE-13776]: https://redpandadata.atlassian.net/browse/CORE-13776?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--9f192576-4262-45e0-8e8f-24f79f12a89f-->
<!--TEST RESULTS-->
<!--21273f2b-e5ce-4fbd-9582-2ffb41b6b299-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74252">build#74252</a></summary>


| test_class           | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                       |
|:---------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests | test_rapid_shadow_link_toggling | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74252#0199e84e-a2f9-448c-9935-091b503e791c | FLAKY         | 18/21    | upstream reliability is '97.59358288770053'. current run reliability is '85.71428571428571'. drift is 11.8793 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling |
| TimeQueryTest        | test_timequery_with_local_gc    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74252#0199e855-3354-4faf-8fe9-740517c12b33 | FLAKY         | 18/21    | upstream reliability is '90.14951627088831'. current run reliability is '85.71428571428571'. drift is 4.43523 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76610">build#76610</a></summary>


| test_class                         | test_method                                | test_arguments                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                               |
|:-----------------------------------|:-------------------------------------------|:-----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterLinkingTopicSyncingWithMtls | test_topic_syncing_include_exclude         | null                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76610#019a9a71-e6a5-46cf-862b-b42b085613a2 | FLAKY         | 19/21    | upstream reliability is '98.52941176470588'. current run reliability is '90.47619047619048'. drift is 8.05322 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithMtls&test_method=test_topic_syncing_include_exclude        |
| LogCompactionTxRemovalTest         | test_tx_control_batch_removal              | null                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76610#019a9a71-e6a3-4551-ab11-20542714c9dd | FLAKY         | 4/21     | upstream reliability is '68.70588235294117'. current run reliability is '19.047619047619047'. drift is 49.65826 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal                     |
| LogCompactionTxRemovalUpgradeTest  | test_tx_control_batch_removal_with_upgrade | {"test_case_name": "Mixed aborts and commits"} | integration | https://buildkite.com/redpanda/redpanda/builds/76610#019a9a71-e6a8-4a50-b140-13dfd21d4558 | FLAKY         | 15/21    | upstream reliability is '88.6986301369863'. current run reliability is '71.42857142857143'. drift is 17.27006 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| RedpandaMixedTestSelfTest          | test_metrics                               | null                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76610#019a9a6a-bb3a-45c4-adf2-a23567a96035 | FLAKY         | 11/21    | upstream reliability is '77.34138972809667'. current run reliability is '52.38095238095239'. drift is 24.96044 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaMixedTestSelfTest&test_method=test_metrics                                       |
</details>
<!--/body-->

dotnwat commented:
/ci-repeat 1
dt-repeat=10
skip-units
skip-redpanda-build
skip-rebase
debug
tests/rptest/tests/timequery_test.py::TimeQueryTest.test_timequery_with_local_gc

dotnwat commented:
/ci-repeat 1
dt-repeat=10
skip-units
skip-redpanda-build
skip-rebase
debug
tests/rptest/tests/log_segment_ms_test.py

dotnwat commented:
/ci-repeat 1
dt-repeat=100
skip-units
skip-redpanda-build
skip-rebase
debug
tests/rptest/tests/log_segment_ms_test.py

dotnwat commented:
/ci-repeat 1
dt-repeat=100
skip-units
skip-redpanda-build
skip-rebase
debug
tests/rptest/tests/log_segment_ms_test.py

dotnwat commented:
/ci-repeat 1
dt-repeat=100
skip-units
debug
tests/rptest/tests/log_segment_ms_test.py

dotnwat commented:
/ci-repeat 1
dt-repeat=10
skip-units
debug
tests/rptest/tests/log_segment_ms_test.py

dotnwat commented:
/ci-repeat 1
dt-repeat=100
skip-units
skip-redpanda-build
debug
tests/rptest/tests/log_segment_ms_test.py

dotnwat commented:
/ci-repeat 1
dt-repeat=100
skip-units
debug
tests/rptest/tests/log_segment_ms_test.py

dotnwat commented:
/ci-repeat 1
dt-repeat=10
skip-units
debug
tests/rptest/tests/log_segment_ms_test.py

dotnwat committed:
test: handle rpk describe group before consumer offsets topic

We've seen describe group race with the creation of consumer offsets
topic. When this happens rpk describe will succeed, but it omits output
that depends on an accurate partition count for the consumer offsets
topic. This results in an altered output which the parsing logic doesn't
handle. Rather than making the parser smarter wait for the topic to
exist which should happen as a side effect and then retry.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
test: support listing internal topics with rpk tool

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

WillemKauf commented on tests/rptest/clients/rpk.py:
```diff
@@ -1054,6 +1054,40 @@ def try_describe_group(group):
 
+        def try_describe_group(group):
+            try:
+                out = describe_group(group)
```
Could de-duplicate the code here by just waiting for `__consumer_offsets` unconditionally before calling into `describe_group()` but I don't have strong feelings about you changing that for the PR.

dotnwat replied:
> waiting for __consumer_offsets unconditionally

right, but it's unusual for the topic to not exist and once it exists (if the race occurs) then it will never not exist after that for the remainder of the test. i'm not sure removing the code duplication justifies the overhead.

WillemKauf commented:
Thanks for the detailed cover letter!

BenPope commented:
/backport v25.3.x
